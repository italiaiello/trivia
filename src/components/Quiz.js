import React, { useState } from 'react'
import { useDataFetch } from '../hooks/displayQuestions'

const Quiz = ({ numQuestions, category, difficulty, type }) => {
    
    const [isLoading, questions] = useDataFetch(`https://opentdb.com/api.php?amount=10`)
                                                

    const [index, setIndex] = useState(0)

    if (questions.length) {
        questions[index].question = questions[index].question.replace(/&quot;/g,'"')
        questions[index].question = questions[index].question.replace(/&#039;/g,"'")
    }
    
    return (
       <section className="quiz">
           {
               isLoading ?
               <h2>Loading Questions</h2>
               :
                <article>
                    {   
                        questions.length &&
                        <p>{questions[index].question}</p>
                    }
                    
                    <button onClick={() => setIndex(index + 1)}>Next Question</button>
                </article>
                
               
               
           }
       </section>
   )
}

export default Quiz

// ${category !== 'Any' && `&category=${category}`}
// ${difficulty !== 'Any' && `&difficulty=${difficulty}`}
// ${type !== 'Any' && `&type=${type}`}