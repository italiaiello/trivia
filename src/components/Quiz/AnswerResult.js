import React from 'react'
import CorrectIcon from '../../images/correct.svg'
import WrongIcon from '../../images/wrong.svg'

const AnswerResult = ({     isAnswerCorrect, 
                            nextQuestion, 
                            setIsAnswerCorrect, 
                            chosenAnswer, 
                            correctAnswer 
                }) => {

    const goToNextQuestion = (e) => {
        nextQuestion()
        setIsAnswerCorrect('unanswered')
    }

    return (
        <article className={isAnswerCorrect ? "answerResult correct" : "answerResult wrong"}>
            {
                isAnswerCorrect
                ?
                <article className="resultIconContainer">
                    <figure className="resultIcon">
                        <img src={CorrectIcon} alt="Correct answer icon" />
                    </figure>
                    <p>{`${chosenAnswer} is correct, well done!`}</p>
                </article>
                :
                <article className="resultIconContainer">
                    <figure className="resultIcon">
                        <img src={WrongIcon} alt="Wrong answer icon" />
                    </figure>
                    <p>{`Sorry, ${chosenAnswer} is incorrect.`}</p>
                    <p>{`The correct answer was ${correctAnswer}`}</p>
                </article>
            }
            <button onClick={goToNextQuestion}>Next Question</button>
        </article>
    )
}

export default AnswerResult