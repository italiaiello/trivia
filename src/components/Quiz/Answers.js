import React, { useState } from 'react'
import AnswerResult from './AnswerResult'

const Answers = ({ incorrectAnswers, correctAnswer, type, nextQuestion }) => {

    const answers = ['1', '2', '3', '4']

    const getRandomInt = (max) => {
        return Math.floor(Math.random() * Math.floor(max));
    }

    if (type === 'multiple') {
        const randomIndex = getRandomInt(4)
        answers[randomIndex] = correctAnswer
        let incorrectIndex = 0;
        for (let i = 0; i < answers.length; i++) {
            if (answers[i] !== correctAnswer) {
                answers[i] = incorrectAnswers[incorrectIndex]
                incorrectIndex++
            }
        }
    }

    const [isAnswerCorrect, setIsAnswerCorrect] = useState('unanswered')

    const checkAnswer = e => {
        if (e.target.textContent === correctAnswer) {
            setIsAnswerCorrect(true)
        } else {
            setIsAnswerCorrect(false)
        }
    }

    return(
        <article>
            {
                type === 'boolean' ?
                <article>
                    <button onClick={checkAnswer}>True</button>
                    <button onClick={checkAnswer}>False</button>
                </article>
                :
                (
                    answers.map((answer, i) => {
                        return (
                            <article key={i}>
                                <button onClick={checkAnswer}>{answer}</button>
                            </article>
                        )
                    })
                )
            }
            {
                isAnswerCorrect !== 'unanswered' &&
                <AnswerResult isAnswerCorrect ={isAnswerCorrect} 
                                nextQuestion={nextQuestion} 
                                setIsAnswerCorrect={setIsAnswerCorrect}
                />
            }
        </article>
    )
}

export default Answers

