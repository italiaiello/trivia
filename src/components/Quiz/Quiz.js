import React, { useState } from 'react'
import { useDataFetch } from '../../hooks/displayQuestions'
import Answers from './Answers'
import QuizFinished from './QuizFinished'

const Quiz = ({ numQuestions, category, difficulty, type }) => {
    
    const [isLoading, questions] = useDataFetch(`https://opentdb.com/api.php?amount=${numQuestions}${category !== 'Any' ? `&category=${category}` : ""}${difficulty !== 'Any' ? `&difficulty=${difficulty}` : ""}${type !== 'Any' ? `&type=${type}` : ""}`)                                   
    const [index, setIndex] = useState(0)
    const [questionsRemaining, setQuestionsRemaining] = useState(numQuestions)
    const [numCorrectAnswers, setNumCorrectAnswers] = useState(0)
    const [isQuizFinished, setIsQuizFinished] = useState(false)

    const nextQuestion = () => {
        if (index < questions.length - 1) {
            setIndex(index + 1)
            setQuestionsRemaining(questionsRemaining - 1)
        } else {
            setIsQuizFinished(true)
        }
    }

    if (questions.length) {
        questions[index].question = questions[index].question.replace(/&quot;/g,'"')
        questions[index].question = questions[index].question.replace(/&#039;/g,"'")
        questions[index].question = questions[index].question.replace(/&eacute;/g,"é")
        questions[index].question = questions[index].question.replace(/&Aacute;/g,"Á")
    }
    
    return (
       <section className="quiz">
           {
               isLoading ?
               <h2>Loading Questions</h2>
               :
               (
                    isQuizFinished ?
                    <QuizFinished   numQuestions={numQuestions} 
                                    numCorrectAnswers={numCorrectAnswers} 
                    />
                    :
                    (   
                        questions.length &&
                        <article className="questionContainer">
                            <p>{`Questions Remaining: ${questionsRemaining}`}</p>
                            <p>{`Correct Answers: ${numCorrectAnswers}`}</p>
                            <p>{questions[index].question}</p>
                            <article className="answers">
                                <Answers    incorrectAnswers={questions[index].incorrect_answers} 
                                            correctAnswer={questions[index].correct_answer}
                                            type={questions[index].type}
                                            nextQuestion={nextQuestion}
                                            numCorrectAnswers={numCorrectAnswers}
                                            setNumCorrectAnswers={setNumCorrectAnswers}
                                />
                            </article>
                        </article>
                    )
               )
                
           }
       </section>
   )
}

export default Quiz