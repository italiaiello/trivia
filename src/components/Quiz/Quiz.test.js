import React from 'react'
import { shallow } from 'enzyme'
import Quiz from './Quiz'

describe("Quiz.js", () => {
    it('checks if it Quiz.js renders properly', () => {
        expect(shallow(<Quiz/>)).toMatchSnapshot()
    })
})


