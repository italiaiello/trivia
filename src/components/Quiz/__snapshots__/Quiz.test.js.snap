// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz.js checks if it Quiz.js renders properly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "quiz",
    },
    "children": Array [
      Object {
        "data": "0",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "section",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
